// @agconnect/database-server/dist/exception/CloudDBErrorCode.js
"use strict";
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2020-2020. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudDBErrorCode = void 0;
/**
 * Cloud DB error Code
 */
var CloudDBErrorCode;
(function (CloudDBErrorCode) {
    CloudDBErrorCode["VALUE_IS_NULL"] = "2001";
    CloudDBErrorCode["DATA_SIZE_IS_OVERFLOW"] = "2003";
    CloudDBErrorCode["QUERY_CONDITION_IS_NULL"] = "2006";
    CloudDBErrorCode["QUERY_INPUT_FIELD_NAME_IS_NULL"] = "2011";
    CloudDBErrorCode["QUERY_INPUT_FIELD_NAME_DO_NOT_EXIST"] = "2012";
    CloudDBErrorCode["INPUT_LIMIT_OR_OFFSET_LESS_THAN_ZERO"] = "2014";
    CloudDBErrorCode["QUERY_INPUT_VALUE_IS_NULL"] = "2015";
    CloudDBErrorCode["GET_ACCESS_TOKEN_FAIL"] = "2016";
    CloudDBErrorCode["NETTY_COMMUNICATION_ERROR"] = "2017";
    CloudDBErrorCode["AGGREGATE_QUERY_CLOUDDB_ZONE_IS_NULL"] = "2018";
    CloudDBErrorCode["AGGREGATE_QUERY_FIELD_NAME_IS_NULL"] = "2019";
    CloudDBErrorCode["AGGREGATE_QUERY_FIELD_TYPE_IS_NOT_NUMERIC"] = "2020";
    CloudDBErrorCode["AGGREGATE_QUERY_ENTITY_DO_NOT_HAVE_FIELD"] = "2021";
    CloudDBErrorCode["QUERY_INPUT_QUERY_OBJECT_IS_NULL"] = "2022";
    CloudDBErrorCode["PAGINATE_QUERY_INPUT_IS_MORE_THAN_ONE"] = "2023";
    CloudDBErrorCode["PAGINATE_QUERY_INPUT_IS_ONLY_SUPPORTED"] = "2024";
    CloudDBErrorCode["PAGINATE_EQUAL_TO_NOT_SUPPORT_PRIMARY_KEY"] = "2025";
    CloudDBErrorCode["PAGINATE_ORDER_BY_NOT_SUPPORT_TYPE"] = "2026";
    CloudDBErrorCode["PAGINATE_EQUAL_TO_FIELD_IS_SAME_WITH_ORDER_BY"] = "2027";
    CloudDBErrorCode["FAIL_TO_PARSE_QUERY_DATA"] = "2035";
    CloudDBErrorCode["PAGINATE_NOT_SUPPORT_AGGREGATE"] = "2036";
    CloudDBErrorCode["EQUAL_TO_NOT_BEFORE_ORDER_BY"] = "2038";
    CloudDBErrorCode["PAGINATE_EQUAL_TO_NOT_SUPPORT_TEXT"] = "2039";
    CloudDBErrorCode["ORDER_BY_ONLY_SUPPORT_ONE_DIRECTION"] = "2040";
    CloudDBErrorCode["PAGINATION_ORDER_BY_HAS_DUPLICATE_FIELD"] = "2042";
    CloudDBErrorCode["DO_NOT_SUPPORT_BYTEARRAY"] = "2043";
    CloudDBErrorCode["DO_NOT_SUPPORT_BYTEARRAY_OR_BOOLEAN"] = "2044";
    CloudDBErrorCode["ONLY_SUPPORT_TEXT_AND_STRING"] = "2045";
    CloudDBErrorCode["NUMBER_IS_INVALID"] = "2046";
    CloudDBErrorCode["CLASS_INVALID"] = "2047";
    CloudDBErrorCode["FAIL_TO_GET_CREDENTIAL"] = "2048";
    CloudDBErrorCode["RESPONSE_IS_INVALID"] = "2049";
    CloudDBErrorCode["ENCRYPTED_FIELD_NOT_SUPPORT_QUERY"] = "2050";
    CloudDBErrorCode["USERID_IS_INVALID"] = "2051";
    CloudDBErrorCode["OBJECT_IS_INVALID"] = "2052";
    CloudDBErrorCode["ORDERBY_DO_NOT_SUPPORT_TEXT_OR_BYTEARRAY"] = "2053";
    CloudDBErrorCode["DATE_STRING_INVALID"] = "2055";
    CloudDBErrorCode["AGCONNECT_CLOUDDB_NOT_INITIALIZE"] = "2056";
    CloudDBErrorCode["DATABASE_NAME_INVALID"] = "2057";
    CloudDBErrorCode["ONLY_SUPPORT_ONE_SCHEMA"] = "2058";
    CloudDBErrorCode["INPUT_CONTAIN_ILLEGAL_CHARACTER"] = "2059";
    CloudDBErrorCode["DUPLICATE_FIELD"] = "2060";
    CloudDBErrorCode["FIELD_NOT_EXIST"] = "2061";
    CloudDBErrorCode["INVALID_FIELD_TYPE"] = "2062";
    CloudDBErrorCode["INCREMENTAL_FIELD_ONLY_SUPPORT_NUMERIC"] = "2063";
    CloudDBErrorCode["PRIMARY_KEY_REQUIRED"] = "2064";
    CloudDBErrorCode["CONDITION_EXCEEDS_LIMIT"] = "2065";
    CloudDBErrorCode["INVALID_CONDITION"] = "2066";
    CloudDBErrorCode["PRIMARY_KEY_ENCRYPTED_FIELD_CANNOT_BE_UPDATED"] = "2067";
    CloudDBErrorCode["OPERATOR_MISMATCH_CONSTRAINT"] = "2068";
    /**
     * invalid CloudDB URL.
     */
    CloudDBErrorCode["URL_INVALID"] = "3007001";
    /**
     * invalid object type name.
     */
    CloudDBErrorCode["TABLE_NAME_INVALID"] = "3007003";
    /**
     * authentication failed.
     */
    CloudDBErrorCode["AUTHENTICATION_ERROR"] = "3007004";
    /**
     * the request type is not supported.
     */
    CloudDBErrorCode["HTTPS_METHOD_INVALID"] = "3007005";
    /**
     * invalid JSON.
     */
    CloudDBErrorCode["ERROR_CODE_JSON_INVALID"] = "3007006";
    /**
     * Data invalid, include data type mismatch and data overflow
     * User modify the data.
     */
    CloudDBErrorCode["ERROR_CODE_DATA_INVALID"] = "3007007";
    /**
     * Parameter invalid
     * User check the parameter corresponding to his operation.
     */
    CloudDBErrorCode["PARAMETER_INVALID"] = "3007008";
    /**
     * too many requests. The system is too busy to process them.
     */
    CloudDBErrorCode["TOO_MANY_REQUESTS"] = "3007009";
    /**
     * insufficient service quota.
     */
    CloudDBErrorCode["QUOTA_NOT_ENOUGH"] = "3007014";
    /**
     * service arrears.
     */
    CloudDBErrorCode["CHARGE_NOT_ENOUGH"] = "3007015";
    /**
     * system error.
     */
    CloudDBErrorCode["SYSTEM_ERROR"] = "3007999";
    /**
     * the number of input data records is 0.
     */
    CloudDBErrorCode["ERROR_CODE_DATA_SIZE_ZERO"] = "3037001";
    /**
     * Data does not conclude primary key
     * User need to modify the data
     */
    CloudDBErrorCode["ERROR_CODE_DATA_NOT_CONCLUDE_PRIMARY_KEY"] = "3037003";
    /**
     * input data field does not match the object type definition.
     */
    CloudDBErrorCode["ERROR_CODE_DATA_MISMATCH_WITH_SCHEMA"] = "3037004";
    /**
     * value of the field cannot be null.
     */
    CloudDBErrorCode["ERROR_CODE_FIELD_EMPTY_WHEN_NOT_NULL"] = "3037005";
    /**
     * single piece of inserted data volume exceeds the upper limit.
     */
    CloudDBErrorCode["ERROR_CODE_DATA_SIZE_EXCEED"] = "3037006";
    /**
     * inserted data is not of the array type.
     */
    CloudDBErrorCode["ERROR_CODE_DATA_NOT_IN_ARRAY"] = "3037007";
    /**
     * invalid query condition.
     * Adjust the query condition.
     */
    CloudDBErrorCode["ERROR_CODE_NOT_CONTAIN_QUERY_CONDITION"] = "3037009";
    /**
     * encrypted fields cannot be used as query conditions.
     */
    CloudDBErrorCode["ERROR_CODE_ENCRYPT_FIELD_CAN_NOT_BE_QUERY_CONDITION"] = "3037010";
    /**
     * queried field does not exist.
     */
    CloudDBErrorCode["ERROR_CODE_FIELD_NAME_NOT_EXIST"] = "3037011";
    /**
     * not support the queried data type under {0} condition.
     */
    CloudDBErrorCode["NOT_SUPPORT_QUERY_DATA_TYPE"] = "3037012";
    /**
     * invalid field name or value.
     */
    CloudDBErrorCode["ERROR_CODE_FIELD_NAME_OR_VALUE_NULL"] = "3037013";
    /**
     * the value range queried by the In condition is not in the array.
     */
    CloudDBErrorCode["ERROR_CODE_IN_QUERY_CONDITION_VALUE_NOT_ARRAY"] = "3037014";
    /**
     * the array size of the value range queried by the In condition is 0.
     */
    CloudDBErrorCode["ERROR_CODE_IN_QUERY_CONDITION_VALUE_ARRAY_SIZE_ZERO"] = "3037015";
    /**
     * invalid OrderBy value.
     */
    CloudDBErrorCode["ERROR_CODE_ORDER_BY_DATA_INVALID"] = "3037016";
    /**
     * the query condition is not supported.
     */
    CloudDBErrorCode["ERROR_CODE_NOT_SUPPORT_QUERY"] = "3037017";
    /**
     * invalid Limit query condition.
     */
    CloudDBErrorCode["ERROR_CODE_LIMIT_DATA_INVALID"] = "3037018";
    /**
     * the function is not supported.
     */
    CloudDBErrorCode["ERROR_CODE_FUNCTION_IS_NOT_SUPPORT"] = "3037019";
    /**
     * the total number of data records to be inserted or deleted is greater than 1000.
     */
    CloudDBErrorCode["OPERATE_DATA_OVERSIZE"] = "3037020";
    /**
     * the query conditions GroupBy and OrderBy cannot coexist.
     */
    CloudDBErrorCode["ERROR_CODE_ORDER_GROUP_NOT_EXIST"] = "3037023";
    /**
     * if the GroupBy query condition exists, the query field (non- aggregate query) must exist in GroupBy.
     */
    CloudDBErrorCode["ERROR_CODE_SELECT_GROUP_FIELD_SAME"] = "3037024";
    /**
     * the value of GroupBy must contain an index.
     */
    CloudDBErrorCode["ERROR_CODE_GROUP_BY_FIELD_MUST_INDEX"] = "3037025";
    /**
     * the EqualTo pagination query condition does not support primary keys.
     */
    CloudDBErrorCode["ERROR_CODE_PAGINATION_QUERY_EQUAL_TO_NOT_SUPPORT_PRIMARY_KEY"] = "3037026";
    /**
     * only one pagination query condition is supported.
     */
    CloudDBErrorCode["ERROR_CODE_PAGINATION_QUERY_CONDITION_NUM_ERROR"] = "3037027";
    /**
     * the input object of pagination query condition is null.
     */
    CloudDBErrorCode["ERROR_CODE_PAGINATION_QUERY_OBJECT_IS_NULL"] = "3037028";
    /**
     * The query condition does not match the index.
     * User need to modify the query condition.
     */
    CloudDBErrorCode["ERROR_CODE_PAGINATION_QUERY_FAILED_FOR_NO_INDEX"] = "3037029";
    /**
     * the OrderBy pagination query condition does not support the Text, Boolean, or ByteArray type.
     */
    CloudDBErrorCode["ERROR_CODE_PAGINATION_QUERY_ORDER_BY_NOT_SUPPORT_TYPE"] = "3037031";
    /**
     * the EqualTo pagination query condition does not support the Text type.
     */
    CloudDBErrorCode["ERROR_CODE_PAGINATION_QUERY_EQUAL_TO_SUPPORT_TYPE"] = "3037032";
    /**
     * the sequence of the OrderBy query condition must be the same.
     */
    CloudDBErrorCode["ERROR_CODE_PAGINATION_QUERY_ORDER_BY_DIRECTION_CONSISTENT"] = "3037033";
    /**
     * the pagination query condition must be after the range query condition and before the Limit query condition.
     */
    CloudDBErrorCode["ERROR_CODE_PAGINATION_QUERY_SEQUENCE"] = "3037034";
    /**
     * The transaction operation sequence is incorrect.
     * User need to modify the query condition.
     */
    CloudDBErrorCode["ERROR_CODE_TRANSACTION_SEQUENCE_INCORRECT"] = "3037035";
    /**
     * the OrderBy pagination query condition does not support duplicate field names.
     */
    CloudDBErrorCode["CLOUD_PAGINATION_ORDER_BY_HAS_DUPLICATE_FIELD"] = "3037036";
    /**
     * the Select statement does not support duplicate field names or aliases.
     */
    CloudDBErrorCode["CLOUD_SELECT_QUERY_NOT_REPEAT"] = "3037037";
    /**
     * MIME type error,only support json/octet-stream.
     */
    CloudDBErrorCode["NOT_SUPPORT_DATA_TYPE"] = "3037038";
    /**
     * updated field does not exist.
     */
    CloudDBErrorCode["ERROR_CODE_FIELD_NAME_NOT_EXIST_IN_UPDATE"] = "3037039";
    /**
     * encrypted fields cannot be updated.
     */
    CloudDBErrorCode["ERROR_CODE_ENCRYPTED_FIELDS_CANNOT_BE_UPDATED"] = "3037040";
    /**
     * schema exceeds the limit in the where condition.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_EXCEEDS_LIMIT_IN_WHERE_CONDITION"] = "3037041";
    /**
     * invalid increment or update.
     */
    CloudDBErrorCode["ERROR_CODE_INVALID_INCREMENT_OR_UPDATE"] = "3037042";
    /**
     * user input data value field type can not increment.
     */
    CloudDBErrorCode["ERROR_CODE_FIELD_TYPE_CANNOT_INCREMENT"] = "3037043";
    /**
     * the operator constraint of object exceeds limit.
     */
    CloudDBErrorCode["ERROR_CODE_TOO_MANY_QUERY_CONDITION"] = "3037044";
    /**
     * the trigger capacity is full. Release the capacity and try again.
     */
    CloudDBErrorCode["LISTENER_TOO_MANY"] = "3047001";
    /**
     * exported data size is too large.
     */
    CloudDBErrorCode["ERROR_CODE_DATA_EXPORT_TOO_MANY"] = "4000009";
    /**
     * invalid AppGallery Connect gateway.
     */
    CloudDBErrorCode["AGC_CLOUD_GATEWAY_INVALID"] = "4007014";
    /**
     * CloudDB service is not enabled.
     */
    CloudDBErrorCode["ERROR_CODE_APP_CREATION_FAILURE"] = "4020001";
    /**
     * number of indexes exceeds the specification limitations {0}.
     */
    CloudDBErrorCode["ERROR_CODE_INDEX_NUMBER_EXCEED"] = "4021001";
    /**
     * number of fields in the index exceeds the specification limitations {0}.
     */
    CloudDBErrorCode["ERROR_CODE_INDEX_FIELD_NUMBER_EXCEED"] = "4021002";
    /**
     * invalid index name.
     */
    CloudDBErrorCode["ERROR_CODE_INDEX_NAME_INVALID"] = "4021003";
    /**
     * duplicate fields in an index.
     */
    CloudDBErrorCode["ERROR_CODE_INDEX_FIELDS_REPEAT"] = "4021005";
    /**
     * index field does not exist.
     */
    CloudDBErrorCode["ERROR_CODE_INDEX_FIELD_NOT_EXIST_IN_SCHEMA"] = "4021006";
    /**
     * type of the index field {0} does not meet the specifications.
     */
    CloudDBErrorCode["ERROR_CODE_INDEX_FIELD_NOT_SUPPORT_SOME_TYPE"] = "4021008";
    /**
     * fields of each index are not in the array.
     */
    CloudDBErrorCode["ERROR_CODE_INDEX_LIST_NOT_ARRAY"] = "4021009";
    /**
     * index field is not in the array.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_INDEXES_NOT_ARRAY"] = "4021010";
    /**
     * index is not in the JSON format.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_INDEX_NOT_JSON_TYPE"] = "4021011";
    /**
     * invalid index.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_INDEX_SETTING_INVALID"] = "4021012";
    /**
     * number of primary keys exceeds the specification limitations {0}.
     */
    CloudDBErrorCode["ERROR_CODE_PRIMARY_KEY_NUMBER_EXCEED"] = "4022001";
    /**
     * encrypted field cannot be set in a primary key.
     */
    CloudDBErrorCode["ERROR_CODE_ENCRYPT_FIELD_BE_PRIMARY_KEY"] = "4022002";
    /**
     * type of the primary key field {0} does not meet the specifications.
     */
    CloudDBErrorCode["ERROR_CODE_PRIMARY_KEY_NOT_SUPPORT_SOME_TYPE"] = "4022003";
    /**
     * primary key field is required.
     */
    CloudDBErrorCode["ERROR_CODE_PRIMARY_KEY_SHALL_NOT_NULL"] = "4022004";
    /**
     * default value cannot be set for the primary key.
     */
    CloudDBErrorCode["ERROR_CODE_PRIMARY_KEY_SHALL_NOT_SET_DEFAULT_VALUE"] = "4022005";
    /**
     * type of the non-null field {0} does not meet the specifications.
     */
    CloudDBErrorCode["NOT_NULL_FIELD_NOT_SUPPORT_SOME_TYPE"] = "4023001";
    /**
     * encrypted fields cannot be set to a non-empty value.
     */
    CloudDBErrorCode["ERROR_CODE_ENCRYPT_FIELD_NOT_SET_NOT_NULL"] = "4023002";
    /**
     * default value of type overflows.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_FIELD_DEFAULT_VALUE_INVALID"] = "4024001";
    /**
     * type of the default value field {0} does not meet the specifications.
     */
    CloudDBErrorCode["SCHEMA_DEFAULT_VALUE_NOT_SUPPORT"] = "4024002";
    /**
     * only the default value of a string type can be set to an empty string.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_DEFAULT_VALUE_EMPTY"] = "4024003";
    /**
     * invalid default value.
     */
    CloudDBErrorCode["ERROR_CODE_DEFAULT_VALUE_INVALID"] = "4024004";
    /**
     * default value cannot be set for an encrypted field.
     */
    CloudDBErrorCode["ERROR_CODE_ENCRYPT_FIELD_SET_DEFAULT_VALUE"] = "4024005";
    /**
     * default value is not set for the non-null field.
     */
    CloudDBErrorCode["ERROR_CODE_NOT_SET_DEFAULT_VALUE_AS_NOT_NULL_TRUE"] = "4024006";
    /**
     * permission is not in the array.
     */
    CloudDBErrorCode["ERROR_CODE_RIGHT_NOT_ARRAY_TYPE"] = "4026001";
    /**
     * the permission field is required.
     */
    CloudDBErrorCode["ERROR_CODE_RIGHT_EMPTY"] = "4026002";
    /**
     * role and permission are not in the JSON format.
     */
    CloudDBErrorCode["ERROR_CODE_RIGHT_MEMBER_NOT_JSON_TYPE"] = "4026003";
    /**
     * permission does not match the role.
     */
    CloudDBErrorCode["ERROR_CODE_RIGHT_MISMATCH_WITH_ROLE"] = "4026004";
    /**
     * no permission is set for the object type {0}.
     */
    CloudDBErrorCode["ERROR_CODE_RIGHT_NOT_SETTING"] = "4026005";
    /**
     * permission is set for an object type that does not exist.
     */
    CloudDBErrorCode["ERROR_CODE_RIGHT_SETTING_FOR_NOT_EXIST_SCHEMA"] = "4026006";
    /**
     * unsupported permission is set.
     */
    CloudDBErrorCode["ERROR_CODE_RIGHT_NOT_SUPPORT"] = "4026007";
    /**
     * invalid object type version number.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_VERSION_INVALID"] = "4027003";
    /**
     * number of object type fields exceeds the upper limit {0}.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_FIELD_NUMBER_EXCEED"] = "4027005";
    /**
     * number of fields of the string type exceeds the upper limit {0}.
     */
    CloudDBErrorCode["ERROR_CODE_STRING_TYPE_NUMBER_EXCEED"] = "4027006";
    /**
     * object type is not in the JSON.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMAS_JSON_EMPTY"] = "4027008";
    /**
     * object type member is not in the JSON format.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMAS_MEMBER_NOT_JSON_TYPE"] = "4027009";
    /**
     * input field of the object type is not in the array.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_FIELDS_NOT_ARRAY"] = "4027010";
    /**
     * object type does not contain fields.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_FIELDS_NOT_EXIST"] = "4027011";
    /**
     * field in the object type is not in the JSON format.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_FIELDS_NOT_JSON_TYPE"] = "4027012";
    /**
     * object type field is incomplete.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_FIELD_ATTRIBUTES_EMPTY"] = "4027013";
    /**
     * object type field name does not meet the specifications.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_FIELD_NAME_INVALID"] = "4027014";
    /**
     * incorrect properties related to the Boolean type in the object type field {0}.
     */
    CloudDBErrorCode["SCHEMA_FIELD_ATTRIBUTES_NOT_BOOLEAN"] = "4027015";
    /**
     * primary key is not set for the object type {0}.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_NO_PRIMARY_KEY"] = "4027017";
    /**
     * invalid object type.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_INVALID"] = "4027018";
    /**
     * invalid package name of the exported object type.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_PACKAGE_NAME_INVALID"] = "4027019";
    /**
     * an exception occurred when the object type package is exported.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_PACKAGE_EXPORT_ERROR"] = "4027020";
    /**
     * the data written per second exceeds 10 MB. Adjust the amount of data written.
     */
    CloudDBErrorCode["WRITE_DATA_SIZE_EXCEED_PER_SECOND"] = "4027023";
    /**
     * permissions of everyone or authenticated users cannot be set for object types containing encrypted fields.
     */
    CloudDBErrorCode["ENCRYPT_FIELD_PERMISSION_INVALID"] = "4027024";
    /**
     * a maximum of five fields can be encrypted for an object type.
     */
    CloudDBErrorCode["ENCRYPT_FIELD_NUM_EXCEED"] = "4027025";
    /**
     * fields of the Text and ByteArray types cannot be encrypted.
     */
    CloudDBErrorCode["ENCRYPT_FIELD_TYPE_INVALID"] = "4027026";
    /**
     * invalid parameter for exporting data.
     */
    CloudDBErrorCode["ERROR_CODE_EXPORT_DATA_PARA_INVALID"] = "4030001";
    /**
     * data of object types containing encrypted fields cannot be exported.
     */
    CloudDBErrorCode["ERROR_CODE_EXPORT_ENCRYPT_SCHEMA"] = "4030002";
    /**
     * importing data to object types that contain encrypted fields is not supported.
     */
    CloudDBErrorCode["ERROR_CODE_IMPORT_ENCRYPT_SCHEMA"] = "4030003";
    /**
     * number of downloaded data records is greater than 1.
     */
    CloudDBErrorCode["ERROR_CODE_DATA_SIZE_BIGGER_THAN_ONE"] = "4037001";
    /**
     * query progress does not exist.
     */
    CloudDBErrorCode["ERROR_CODE_PROCESS_ID_INVALID"] = "4038001";
    /**
     * the number of tables exceeds the limit.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_NUMBER_EXCEED"] = "2001005";
    /**
     * the primary key cannot be modified.
     */
    CloudDBErrorCode["PRIMARY_KEY_NOT_SUPPORT_CHANGE"] = "2001006";
    /**
     * permission denied.
     */
    CloudDBErrorCode["ERROR_CODE_PERMISSION_DENIED"] = "2001015";
    /** the capacity of objects exceeds the limit.
     * Adjust the query condition.
     */
    CloudDBErrorCode["ERROR_QUERY_DATA_SIZE_TOO_LARGE"] = "2001017";
    /** the size of object list exceeds the limit.
     * Adjust the query condition.
     */
    CloudDBErrorCode["ERROR_QUERY_DATA_COUNT_TOO_LARGE"] = "2001018";
    /**
     * the size of the data to be written exceeds the limit.
     */
    CloudDBErrorCode["ERROR_CODE_DATA_UPSERT_TOO_LARGE"] = "2001020";
    /**
     * the count of update fields exceeds the limit.
     */
    CloudDBErrorCode["UPDATE_FIELDS_COUNT_EXCEED_LIMIT"] = "2001035";
    /**
     * the count of operators in expression exceeds the limit.
     */
    CloudDBErrorCode["OPERATORS_COUNT_IN_EXPRESSION_EXCEED_LIMIT"] = "2001036";
    /**
     * the original object type has been changed.
     */
    CloudDBErrorCode["SCHEMA_ALREADY_CHANGED"] = "2002005";
    /**
     * Schema not exist
     * User need to check the schema verification
     */
    CloudDBErrorCode["SCHEMA_NOT_EXISTING"] = "2002007";
    /**
     * duplicate field name.
     */
    CloudDBErrorCode["ERROR_CODE_SCHEMA_FIELD_NAME_SAME"] = "2002012";
    /**
     * duplicate index name.
     */
    CloudDBErrorCode["ERROR_CODE_INDEX_NAME_REPEAT"] = "2002018";
    /**
     * invalid parameter
     */
    CloudDBErrorCode["ERROR_CODE_PARAM_INVALID"] = "2002000";
    /**
     * not null field can not be null
     */
    CloudDBErrorCode["ERROR_CODE_NOT_NULL_FIELD_CANNOT_BE_NULL"] = "2002036";
    /**
     * CloudDBZone is not exist
     */
    CloudDBErrorCode["ERROR_CODE_DATABASE_NOT_EXIST"] = "2002037";
    /**
     * the primary key cannot be updated.
     */
    CloudDBErrorCode["PRIMARY_KEY_CANNOT_BE_UPDATED"] = "2002041";
    /**
     * incremental or updated object is not exist.
     */
    CloudDBErrorCode["ERROR_CODE_OBJECT_NOT_EXIST"] = "2002043";
    /**
     * primary key duplicate when insert date
     */
    CloudDBErrorCode["ERROR_CODE_PRIMARY_KEY_DUPLICATE"] = "2002044";
    /**
     * only BYTE, SHORT, INT, LONG, FLOAT, DOUBLE type support increment
     */
    CloudDBErrorCode["ERROR_CODE_FIELD_TYPE_NOT_SUPPORT_INCREMENT"] = "2002045";
    /**
     * the cache plan has been changed.
     */
    CloudDBErrorCode["ERROR_TRANSACTION_FAILED_DATA_CHANGED"] = "2003049";
    /**
     * the value range exceeds the limit.
     */
    CloudDBErrorCode["ERROR_CODE_VALUE_OUT_OF_RANGE"] = "2003053";
    /**
     * the number of index fields exceeds the limit.
     */
    CloudDBErrorCode["ERROR_CODE_IMPORT_DATA_INDEX_FIELD_VALUE_LENGTH_EXCEED"] = "2003094";
})(CloudDBErrorCode = exports.CloudDBErrorCode || (exports.CloudDBErrorCode = {}));
