// @agconnect/database-server/dist/services/CloudDBService.js
"use strict";
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2020-2020. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudDBService = void 0;
var CloudDBErrorCode_1 = require("../exception/CloudDBErrorCode");
var AGConnectCloudDBException_1 = require("../exception/AGConnectCloudDBException");
var ErrorCodeMessage_1 = require("../exception/ErrorCodeMessage");
var HttpsCommunicator_1 = require("../communicator/https/HttpsCommunicator");
var Utils_1 = require("../utils/Utils");
/**
 * The base class of cloudDB service.
 */
var CloudDBService = /** @class */ (function () {
    /**
     * Constructor of CloudDB Service.
     *
     * @param agcClient client from Agc.
     * @throws AGConnectCloudDBException if agcClient is null.
     */
    function CloudDBService(agcClient) {
        this.URL_PREFIX = 'api/clouddb/clouddbservice/';
        if (!agcClient) {
            console.warn(new AGConnectCloudDBException_1.AGConnectCloudDBException(CloudDBErrorCode_1.CloudDBErrorCode.VALUE_IS_NULL).message);
            throw new AGConnectCloudDBException_1.AGConnectCloudDBException(CloudDBErrorCode_1.CloudDBErrorCode.VALUE_IS_NULL);
        }
        this.agcClient = agcClient;
        this.requestSender = new HttpsCommunicator_1.HttpsCommunicator(this.agcClient);
    }
    /**
     * This method is used to send Delete request and parse the response.
     *
     * @param requestUrl the request url.
     * @param data the data need to send.
     * @returns the number of recprds that being deleted.
     * @throws AGConnectCloudDBException if response is invalid.
     */
    CloudDBService.prototype.objectDeleteResponse = function (requestUrl, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var requestId = Utils_1.Utils.getRequestId();
            _this.requestSender.sendDeleteRequest(requestUrl, requestId, data).then(function (resp) {
                _this.checkResponseBody(resp);
                resolve(resp['data']['delNumber']);
            }).catch(function () {
                _this.resendDeleteRequest(resolve, reject, requestUrl, requestId, data);
            });
        });
    };
    CloudDBService.prototype.resendDeleteRequest = function (resolve, reject, requestUrl, requestId, data) {
        var _this = this;
        this.requestSender.sendDeleteRequest(requestUrl, requestId, data, true).then(function (resp) {
            _this.checkResponseBody(resp);
            resolve(resp['data']['delNumber']);
        }).catch(function (err) {
            console.warn("Delete catch error, requestId:" + requestId);
            if (_this.isErrResponseNotValid(err)) {
                reject(err);
                return;
            }
            reject(new AGConnectCloudDBException_1.AGConnectCloudDBException(String(err.response.data.errorCodeV2)));
        });
    };
    /**
     * This method is used to check whether the response body is valid.
     *
     * @param resp response body
     */
    CloudDBService.prototype.checkResponseBody = function (resp) {
        if (resp['status'] === 200 && !resp['data']) {
            console.warn(new AGConnectCloudDBException_1.AGConnectCloudDBException(CloudDBErrorCode_1.CloudDBErrorCode.RESPONSE_IS_INVALID).message);
            throw new AGConnectCloudDBException_1.AGConnectCloudDBException(CloudDBErrorCode_1.CloudDBErrorCode.RESPONSE_IS_INVALID);
        }
    };
    /**
     * This method is used to check whether the error response is valid.
     */
    CloudDBService.prototype.isErrResponseNotValid = function (err) {
        if (err.response == null || err.response.data == null || err.response.data.errorCodeV2 == null
            || !ErrorCodeMessage_1.getErrorMessage(err.response.data.errorCodeV2)) {
            return true;
        }
        console.warn('Get cloud error code =>', err.response.data.errorCodeV2);
        return false;
    };
    return CloudDBService;
}());
exports.CloudDBService = CloudDBService;
